rfcs/0075-declarative-wrappers.md

```diff
@@ -0,0 +1,367 @@
+introduce a new `makeWrapperAuto` setup hook that will take care of this in the
+way described as follows.
+
+As a start, we'll need to think about all packages in Nixpkgs that using them
+requires some environment variables to be set. Every such package will put in
+`$dev/nix-support/wrappers.json` a list of environment variables that are
+"linked" to this package. "linked" means that using this package requires these
+environment variables to be set in runtime.
+
+`makeWrapperAuto` will traverse all `buildInputs` and `propagatedBuildInputs`
+of a derivation, and look for a `wrappers.json` file in these inputs. It will
+collect all the environment variables that need to be set in the resulting
+executables, by merging all of the values of the environment variables in all
+of the inputs' `wrappers.json` files. `wrappers.json` might look like this for
+a given package:
+
+```json
+{
+  "GI_TYPELIB_PATH": [
+    "/nix/...gobject-introspection.../...",
+    "/nix/...librsvg.../...",
+  ],
+  "GIO_EXTRA_MODULES": [
+    "/nix/...dconf.../lib/gio/modules"
+  ],
+  "XDG_DATA_DIRS": [
+    "/nix/...gtk+3.../...",
+    "/nix/...gsettings-desktop-schemas.../..."
+  ],
+  "GDK_PIXBUF_MODULE_FILE": "/nix/...librsvg.../lib/gdk.../loaders.cache",
+}
+```
+
+The information found inside an input's `wrappers.json` will include all the
+information about wrappers found in the input's inputs, and so on. Thus in
+contrast to the [POC Nixpkgs PR](https://github.com/NixOS/nixpkgs/pull/85103)
+and the [original design of the
+RFC](https://github.com/doronbehar/rfcs/blob/60d3825fdd4e6574b7e5d70264445d1c801368c6/rfcs/0075-declarative-wrappers.md#L251),
+prior to [the 1st
+meeting](https://github.com/NixOS/rfcs/pull/75#issuecomment-760942876),
+traversing all the inputs and the inputs' inputs, will not happen during eval
+time and only partly, during build time - every package already built will
+provide it's reverse dependencies all the information they need about
+environment variables of itself and of all of it's inputs and it's inputs'
+inputs.
+
+Most of the work to do will be:
+
+1. Gather information about what environment variables are "linked" to each
+   package, and edit these derivations to include a `wrappers.json` in them.
+   This should be done with `makeWrapperAuto` as well, see (2).
+2. Design the `makeWrapperAuto` shell hook:
+  - It should introduce a shell function (to be called `wrappersInfo`) that
+    will allow piping a JSON string from `builtins.toJSON` and spit a
+    `wrappers.json` that will include both what was piped into it, and the
+    content from the package's various inputs' `wrappers.json` files.
+  - It should make the executables in `$out/bin/` get wrapped according to
+    what's currently in this package's `wrappers.json`, during `fixupPhase`.
+  - The above should be also possible to do manually for executables outside
+    `$out/bin/` with say adding to a derivation a Nix variable:
```

Comment:
```suggestion
    `$out/bin/` by setting `wrapExtraPrograms` on the derivation:
```
Maybe we should also not wrap everything in `$out/bin` indiscriminately though? Maybe we should have a variable specifying _exactly_ which files to wrap, not just which files to wrap in addition to `$out/bin`.